File: dart/lib/src/parser/clause.dart
import 'match_result.dart';
import 'parser.dart';

/// Base class for all grammar clauses.
abstract class Clause {
  const Clause();
  MatchResult match(Parser parser, int pos);
  void checkRuleRefs(Map<String, Clause> grammarMap);
  @override
  String toString() => runtimeType.toString();
}

File: java/src/main/java/com/squirrelparser/clause/Clause.java
package com.squirrelparser.clause;

import java.util.Map;

import com.squirrelparser.parser.MatchResult;
import com.squirrelparser.parser.Parser;

/**
 * Base class for all grammar clauses.
 */
public abstract class Clause {

    /**
     * Match this clause at the given position.
     *
     * @param parser The parser instance
     * @param pos    The position in the input
     * @param bound  Optional bound clause for recovery
     * @return The match result
     */
    public abstract MatchResult match(Parser parser, int pos, Clause bound);

    /**
     * Check that all rule references in this clause are valid.
     *
     * @param grammarMap The map of rule names to clauses
     */
    public abstract void checkRuleRefs(Map<String, Clause> grammarMap);

    @Override
    public String toString() {
        return getClass().getSimpleName();
    }
}
